\documentclass[11pt, a4paper,titlepage]{article}
\usepackage[utf8]{inputenc}
\usepackage{lipsum}
\usepackage{latexsym}
\usepackage{hyperref}
\usepackage[left=2.35cm, right=3.35cm, top=3.35cm, bottom=3.35cm]{geometry}
\usepackage[english]{babel}
\usepackage{graphicx}
\usepackage{titlesec}
\usepackage{tocbibind}

\begin{document}

\include{title}
\setcounter{tocdepth}{3}

\tableofcontents
\newpage

\section{Introduction}

Clustered, Regularly Interspaced, Short Palindromic Repeats (CRISPR)
and CRISPR-associated proteins (Cas proteins) form a complex
bacterial/archaeal immune response system that mitigates foreign DNA
activity. The identification of these systems first occured in 1987 in
\emph{Escherichia coli} \cite{nakata1989unusual}, quickly followed by
other species of bacteria and archae \cite{mojica1995long}, but their
function was not elucidated until 2007 \cite{Barrangou23032007}. This
immune-like system works by capturing short signatures of invading DNA
and inserting them into the genome of the organism in regions known as
CRISPR arrays.

These arrays consist of the captured elements, known as spacers, which
are separated by similarly sized, conserved DNA sequences known as
repeats.  The arrays are expressed and processed into CRISPR RNA
(crRNA). crRNA, together with Cas proteins, form an active complex
that patrols the cell. If an invader DNA element with a similar
signature is encountered by the complex, it will be degraded and its
activity will be prevented \cite{Horvath08012010}.

CRISPR-Cas systems can be categorized into multiple types. Each type
has a different set of characteristics that makes it unique
\cite{makarova2011evolution}.

Previously our team used computational CRISPR detection software like
CRT to detect spacers from all ``reference'' bacterial and archael
genomes on NCBI (about 5000 genomes in total in May 2015). A
subsequent step had been to blast those detected elements against the
NCBI nucleotide database. This revealed the origin of about 7\% of
spacers, outlining the problem of the so-called \emph{biological dark
  matter}. The identified spacers were also searched against the
genomes of the bacteria they originated from. Spacers showing up
outside of the CRISPR region were classified as \emph{hits}. Hits
putatively showed that there was a possibility for given phages to
integrate into bacterial genome despite the presence of putative
spacers against them.

The main objective of this project was to make a large database of
CRISPR elements from all the bacterial/archaeal genomes available in
the NCBI genome database. Hence, removing the limitations we had
before when we focused only on ``reference genomes''. The number of
genomes considered is thus substantially higher, about 54000 at the
time of this project (December 2015). A database of these elements was
made accessible via a web application. The associated website exposes
the CRISPR details of the strains, along with a number of services,
such as blasting against spacer elements, CRISPR array detection, and
elements of data visualization.

\section{Methods and results}

\subsection{Classification of CRISPR-Cas systems}
We attempted to classify CRISPR-Cas systems based on methods proposed
by Makarova et al. Data set from their studies (supplementary document
7) was first processed into a clean subset (without “partial” subtypes
and ambiguous classification) to search for possible classifiers. Loci
were classified as “partial” if they contained neither the full
complement of effector module genes nor cas9 or cpf1. The
classification criteria was polythetic classification as mentioned in
their studies [8].

Two approaches were attempted to tackle the problem, random forest and
Naive Bayes. Random forest showed unpromising results as the total
factor level exceeded computational limit. On the other hand, Naive
Bayes model performed well with the clean dataset with overall
accuracy of 0.98, high sensitivity and specificity to all subtypes
except lower sensitivity to CAS-I-A (0.85) and CAS-III-C (0.93) in the
validation set.

The task proved to be non-trivial and was eventually discontinued due
to time constraint. Challenges encountered in the process include
ambiguous classification, need for a pipeline to detect and annotate
cas genes, and genomes bias etc.

\subsection{Data Processing Pipeline}

The first step of the data processing pipeline was to gather all
possible bacterial and archaeal genomes available from the NCBI genome
database. This task was accomplished using a custom python script
which fetched from the RefSeq ftp site \cite{pruitt2007ncbi} the 54000
prokaryote genomes available (all assembly levels). In practice that
meant downloading the genome files (*.fna). Alongside each genome
file, the corresponding annotation files (*.gff) were also
downloaded. Finally the corresponding metadata available on NCBI for
each genome was captured into JSON files. All the files, representing
about 300 Gb worth of data were stored in a ``one genome, one folder''
fashion on our server.

The second step was to detect all CRISPR arrays (i.e spacers and
repeats) in each genome. For this, the CRT software
\cite{bland2007crispr} was used. In order to treat the the multi-FASTA
files correctly (e.g for genomes containing plasmid sequences, or
multiple contigs) the CRT software was wrapped in a python script and
then applied to the fetched genomes to create the CRT report file
containing detected CRISPR array(s) in each strain. Next, a custom
parser was implemented to extract the spacers and repeats from each
CRT report and structure them into a JSON file.

These CRISPR elements and the metadata fetched previously were used to
populate the relational MySQL database. All spacers available were
extracted from the MySQL database and consolidated into a multi-FASTA
file and a blastable database was made using the formatdb tool from
the NCBI BLAST.

\begin{center}
\includegraphics[scale=0.5]{figures/pipeline.png}
\end{center}

\subsection{Data Visualization}

Row chart shows the log10 number of genome sequences used to populate
CRISPR Exposed for each group of prokaryote. The purpose of showing
this chart is to inform users that the content of the database is
highly biased and this should be taken into account when interpreting
the distribution of repeats’ and spacers’ length. About half of the
total of the sequences originate from firmicutes, proteobacteria and
actinobacteria. This chart could be further improved into scale-stack
row chart in the future to reduce the cognitive load of interpreting
the results.

Histograms of repeats’ and spacers’ length mainly serve the purpose of
quality control of CRISPR elements detected by CRT. Length of direct
repeats should not be deviated too much from length of spacers. Data
set used to plot the histograms was preprocessed to remove false
positive that is not bound by the constraint of parameter settings in
CRT. For instance, the maximum length of spacers for CRISPR detection
was set to 48 nucleotide long but there were about 3.6\% (28663 out of
790903) of total number of spacers detected by CRT is longer than the
length specified. These entries were stripped from the data set to
avoid misleading information.

Parallel coordinates shows that most of the genomes in database has
size around 2-7 million bp long and most of them contains 1-5 CRISPR
arrays. No notable correlation between number of CRISPR array and
genome size was observed.

Future improvement in data visualization is necessary to bring more
insight to users. Additional information which could be interesting to
be visualised are:
\begin{enumerate}
\item Cas genes detected: it is possible that CRISPR arrays detected
  is not associated with any cas genes in near proximity. Such arrays
  may be degenerated or have other functions
  \cite{mandin2007identification}
\item Display the type of CRISPR-Cas systems 
\item Showing the folding of repeats: Kunin et al. clustered putative
  secondary structures of crRNA based on sequence similarity and
  ability to form stable secondary structures
  \cite{kunin2007evolutionary}. Combining information from cas genes
  and repeats may provide insights on function and regulation of
  CRISPR-Cas.
\item Protospacer-adjacent-motifs (PAM) sequences: the selection of
  protospacers from invading DNA was shown to be determined by
  recognition of PAMs which is several nucleotides long and varies
  between different CRISPR-Cas systems
  \cite{mojica2009short,deveau2008phage}
\end{enumerate}

\section{Discussion}


The application and the database attached to it is novel in a couple
of ways. To our knowledge, it is the first time such an extensive
identification of CRISPR elements is conducted. About 54,000 genomes
were used in this analysis. Other projects such as CRISPRdb \cite{grissa2007crisprdb}, and
CRISPI \cite{rousseau2009crispi} only focus on the subset of genomes that have complete
level assembly (4,000 genomes for CRISPRdb, 1,200 for CRISPI). As
such, the spacer database constructed is the largest presently
available online, consisting of 800,000 spacer sequences. We hope it
can be useful to researchers, and can envision a couple of uses of our
dataset: Comparative analysis of CRISPR arrays at species, class, or
phylum level.  Distribution of the origin of spacers in given
bacteriophages, or plasmids.  Phage host prediction.  Future work
could include:

\begin{itemize}
\item CRISPR-Cas systems identification and classification,
\item PAM sequences information
\item Detection of anti-CRISPR genes. 
\end{itemize}

\newpage
\bibliographystyle{ieeetr} 
\bibliography{bib-db}
\end{document}
